### ISSUE TITLE
Edward + TF Slim

### ISSUE THREAD

**Issue by simonkamronn** [blei-lab/edward#245]
In trying to use TF Slim together with Edward I've tried to modify the bayesian regression example using neural networks to parameterise the mean and std. It should be simple but it doesn't seem to train the neural network part so I'm guessing the TF Slim part is missing something. Ideas?

```
import numpy as np
import tensorflow as tf
import tensorflow.contrib.slim as slim

import edward as ed
from edward.stats import norm

class DensityNetwork:
    def neural_network(self, X):
        """mu, sigma = NN(x; theta)"""
        hidden1 = slim.fully_connected(X, 128, scope='q/fc1')
        hidden2 = slim.fully_connected(hidden1, 128, scope='q/fc2')
        self.mus = slim.fully_connected(hidden2, 1, activation_fn=None, scope='q/mu')
        self.sigmas = slim.fully_connected(hidden2, 1, activation_fn=None, scope='q/sigma')
        self.sigmas = tf.exp(tf.maximum(self.sigmas, 1e-9))

    def log_prob(self, xs, zs):
        X, y = xs['X'], xs['y']
        self.neural_network(X)
        result = tf.exp(norm.logpdf(y, self.mus, self.sigmas))
        result = tf.log(tf.reduce_sum(result, 1))
        return tf.reduce_sum(result)

def build_toy_dataset(N, noise_std=0.1):
    X = np.concatenate([np.linspace(0, 2, num=N / 2),
                        np.linspace(6, 8, num=N / 2)])
    y = 5.0 * X + norm.rvs(0, noise_std, size=N)
    X = X.reshape((N, 1))
    return X.astype(np.float32), y.astype(np.float32)

N = 40  # num data points
D = 1  # num features
train_data, train_target = build_toy_dataset(N)
test_data, test_target = build_toy_dataset(N)

X = ed.placeholder(tf.float32, shape=(None, D))
y = ed.placeholder(tf.float32, shape=(None,))
data = {'X': X, 'y': y}

model = DensityNetwork()
inference = ed.MAP([], data, model)
sess = ed.get_session()

with tf.variable_scope("model", reuse=None) as scope:
    inference.initialize()
    
NEPOCH = 1000
train_loss = np.zeros(NEPOCH)
test_loss = np.zeros(NEPOCH)
for i in range(NEPOCH):
    _, train_loss[i] = sess.run([inference.train, inference.loss], 
                                 feed_dict={X: train_data, y: train_target})
    test_loss[i] = sess.run(inference.loss, feed_dict={X: test_data, y: test_target})

pred_means, pred_std = sess.run([model.mus, model.sigmas], feed_dict={X: test_data})
```

**Comment by dustinvtran** [comment]
It runs for me (most of the time). Other times it goes to `NaN`s which I presume has to do with instability of the updates. Setting the seed and tweaking the architecture (e.g., with `tf.nn.softplus` parameterization over `tf.exp`) helps.

fyi, this runs on the latest development version
```python
#!/usr/bin/env python
"""Mixture density network using maximum likelihood.
"""
import numpy as np
import tensorflow as tf
import tensorflow.contrib.slim as slim

import edward as ed
from edward.stats import norm


class DensityNetwork:
  def neural_network(self, X):
    """mu, sigma = NN(x; theta)"""
    hidden1 = slim.fully_connected(X, 128, scope='q/fc1')
    hidden2 = slim.fully_connected(hidden1, 128, scope='q/fc2')
    self.mus = slim.fully_connected(hidden2, 1, activation_fn=None, scope='q/mu')
    self.sigmas = slim.fully_connected(hidden2, 1, activation_fn=None, scope='q/sigma')
    self.sigmas = tf.nn.softplus(tf.maximum(self.sigmas, 1e-9))

  def log_prob(self, xs, zs):
    X, y = xs['X'], xs['y']
    self.neural_network(X)
    result = tf.exp(norm.logpdf(y, self.mus, self.sigmas))
    result = tf.log(tf.reduce_sum(result, 1))
    return tf.reduce_sum(result)


def build_toy_dataset(N, noise_std=0.1):
  X = np.concatenate([np.linspace(0, 2, num=N / 2),
            np.linspace(6, 8, num=N / 2)])
  y = 5.0 * X + norm.rvs(0, noise_std, size=N)
  X = X.reshape((N, 1))
  return X.astype(np.float32), y.astype(np.float32)


ed.set_seed(42)

N = 40  # num data points
D = 1  # num features
train_data, train_target = build_toy_dataset(N)
test_data, test_target = build_toy_dataset(N)

X = ed.placeholder(tf.float32, shape=(None, D))
y = ed.placeholder(tf.float32, shape=(None,))
data = {'X': X, 'y': y}

model = DensityNetwork()
inference = ed.MAP([], data, model)
sess = ed.get_session()

with tf.variable_scope("model", reuse=None) as scope:
  inference.initialize()

init = tf.initialize_all_variables()
init.run()

NEPOCH = 1000
train_loss = np.zeros(NEPOCH)
test_loss = np.zeros(NEPOCH)
for i in range(NEPOCH):
  info_dict = inference.update(feed_dict={X: train_data, y: train_target})
  train_loss[i] = info_dict['loss']
  test_loss[i] = sess.run(inference.loss, feed_dict={X: test_data, y: test_target})
  if i % inference.n_print == 0:
    print("Train Loss: {:0.3f}, Test Loss: {:0.3f}".format(train_loss[i],
                                                           test_loss[i]))

pred_means, pred_std = sess.run([model.mus, model.sigmas], feed_dict={X: test_data})
```

**Comment by dustinvtran** [comment]
To isolate the problem of replacing Keras with TensorFlow Slim, I wrote a script with TensorFlow slim here: [`examples/tf_mixture_density_network_slim.py`](https://github.com/blei-lab/edward/blob/7a486d557979507f028f08b412e10e2c817c6d9e/examples/tf_mixture_density_network_slim.py). It is the same data, model, and inference as [`examples/tf_mixture_density_network.py`](https://github.com/blei-lab/edward/blob/7a486d557979507f028f08b412e10e2c817c6d9e/examples/tf_mixture_density_network.py), which uses Keras.

Up to initialization and running the algorithm shorter or longer, the two reach the same optima.

**Comment by simonkamronn** [comment]
Awesome, thanks
